### DISCUSSION TITLE
[BUG] No biomass flux when another objective is added for FBA analysis

### DISCUSSION THREAD

**Discussion by hites77** [opencobra/cobrapy#1387] [Q&A] [answered]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

I was trying to get glycerol production along with biomass growth in my fungal model.
When objective is set separately, it shows biomass and glycerol production. When I set the weightage for both the objective simultaneously, the FBA solution shows some value along with glycerol production, but there was zero flux associated with biomass growth. following are the FBA results of 4 different run with different outcomes:

Biomass as objective (Weight:1)   =  0.0824 mmol/gDW/hr , biomass = 0.0824; glycerol = 0
Biomass as objective : 10 =  0.8242 mmol/gDW/hr, biomass = 0.0824; glycerol = 0
Biomass: 10, Glycerol: 2    =  3.1179 mmol/gDW/hr , biomass = 0; glycerol = 1.558927668
Biomass: 10, Glycerol: 5    =  7.7946 mmol/gDW/hr , biomass = 0; glycerol = 1.558927668

how to solve this issue ?

I even tried it in COBRA toolbox in MATLAB but similar result

### Code sample

Code run:

```python
```

Traceback:

```text
```

### Environment

<details>

</details>

**Comment by Midnighter** [comment]
Please share the code you used to prepare the different objectives and solutions.

**Comment by hites77** [comment]
Here is a simple version of the code where I set 0.8 weight for biomass and 0.2 for glycerol production.

    # Defining reactions representing for objectives
    reaction_1 = model.reactions.get_by_id("r_2111")
    reaction_2 = model.reactions.get_by_id("r_1808")
    
    # Set weights for each objective
    weight_1 = 0.8  # Higher weight for objective 1
    weight_2 = 0.2  # Lower weight for objective 2
    
    model.objective = {reaction_1: weight_1, reaction_2: weight_2}
    
    # The upper bound set at 1000 to get the actual optimal value
    model.reactions.get_by_id("r_1808").upper_bound = 1000.0

    solution = model.optimize()
    print (model.summary(solution=solution))
    print (model.reactions.r_2111.flux)
    

Output:

    Objective
=========
0.2 r_1808 + 0.8 r_2111 = 0.3117855336368235

Uptake
------
Metabolite Reaction   Flux  C-Number  C-Flux
    s_0565   r_1714      1         6 100.00%
    s_1277   r_1992 0.5438         0   0.00%
    s_0805   r_2100 0.2357         0   0.00%

Secretion
---------
Metabolite Reaction   Flux  C-Number C-Flux
    s_0458   r_1672 -1.323         1 22.05%
    s_0766   r_1808 -1.559         3 77.95%

0.0

here r_1672 is for carbon dioxide

**Comment by Midnighter** [comment]
Maybe what you're looking for is a [production envelope](https://cobrapy.readthedocs.io/en/latest/phenotype_phase_plane.html)? It's not surprising to have zero biomass in the way that the problem is set up.

**Reply by hites77** [reply]
In phenotype phase it evaluate the uptake rate of two different metabolite. Whereas here i checking production of biomass and glycerol when the media constituent remain the same. And i didn't got when you say "not surprising to have zero biomass in the way that the problem is set up." can you elaborate it please

**Reply by Midnighter** [reply]
You're not limited to varying uptake, you can absolutely look at the trade off between growth and production flux, too.

@cdiener has explained why it's not surprising. The biomass flux requires the production of a lot of different components, some of which are expensive to make. So in a (weighted) sum as you have formulated, the maximal objective will almost always have a zero biomass entry, since your production flux can be much higher that way.

**Comment by cdiener** [comment]
Agreed. It's kind of expected. In general if you maximize a sum of 2 individual objectives there is no guarantee both are non-zero. In your case your glycerol flux is larger than the biomass flux so it will dominate the objective. Here is also a [good overview](https://en.wikipedia.org/wiki/Multi-objective_optimization) of the issues with multi-objective optimization.

**Reply by hites77** [reply]
Okay, but from logical point it seems that model is just accumulating glycerol without biomass growth. So, to get optimal solution either the fluxes in the model need optimization or another LP solver can solve it better ?

**Reply by cdiener** [reply, marked as answer]
Yep because there is probably a tradeoff. If you produce biomass you have to produce less glycerol or vice versa. What you can do is serial optimization. So maximize biomass, than constrain it to 95% of the maximum for example and follow that with glycerol maximization. That would probably be more biologically relevant.

**Reply by hites77** [reply]
okay, thanks for the support
